drakrun/drakrun/storage.py

```diff
@@ -308,9 +310,204 @@ def import_vm0(self, path: str):
+        except subprocess.CalledProcessError as exc:
+            logging.debug(exc.output)
+            raise RuntimeError("Couldn't create snapshot")
+
+    def get_vm_disk_path(self, vm_id: int) -> str:
+        """ Returns disk path for given VM as defined by XL configuration """
+        return f"phy:/dev/{self.lvm_volume_group}/vm-{vm_id},hda,w"
+
+    def rollback_vm_storage(self, vm_id: int):
+        """ Rolls back changes and prepares fresh storage for new run of this VM """
+        vm_id_vol = os.path.join("/dev", f"{self.lvm_volume_group}", f"vm-{vm_id}")
+
+        if vm_id == 0:
+            raise Exception("vm-0 should not be rollbacked")
+
+        logging.info(f"Rolling back changes to vm-{vm_id} disk")
+        if os.path.exists(vm_id_vol):
+            try:
+                subprocess.check_output(
+                    [
+                        "lvremove",
+                        "-v",
+                        "-y",
+                        # "--noudevsync",
+                        f"{self.lvm_volume_group}/vm-{vm_id}"
+                    ],
+                    stderr=subprocess.STDOUT,
+                )
+            except subprocess.CalledProcessError as exc:
+                logging.debug(exc.output)
+                raise RuntimeError(f"Failed to discard previous logical volume {self.lvm_volume_group}/vm-{vm_id}")
+
+        try:
+            subprocess.check_output(
+                [
+                    "lvcreate",
+                    "-s",
+                    "-L", self.snapshot_disksize,
+                    "-n", f"vm-{vm_id}",
+                    f"{self.lvm_volume_group}/vm-0"
+                ],
+                stderr=subprocess.STDOUT
+            )
+        except subprocess.CalledProcessError as exc:
+            logging.debug(exc.output)
+            raise RuntimeError("Couldn't rollback disk")
+
+    def get_vm0_snapshot_time(self):
+        """ Get UNIX timestamp of when vm-0 snapshot was last modified """
+
+        dir_path = '/etc/lvm/archive'
+        required_file_names = filter(lambda f: self.lvm_volume_group in f, os.listdir(dir_path))
+        file_name = sorted(required_file_names, key=lambda f: os.stat(os.path.join(dir_path, f)).st_mtime)[-1]  # last updated file
+
+        # Ugly code, any better alternatives?
+        with open(os.path.join(dir_path, file_name), 'r') as f:
+            lines = f.readlines()
+            flag = [False, False]
+            for i in lines:
+                i = i.strip()
+                if 'logical_volumes {' == i:
+                    flag[0] = True
+                if 'vm-0-snap {' == i:
+                    flag[1] = True
+                if flag[0] and flag[1]:
+                    try:
+                        return re.search(r'^creation_time = (\d*)$', i).group(1)
+                    except Exception as e:
+                        raise Exception("Creation time could not be found") from e
+                    else:
+                        break
```

Comment:
```suggestion
    def get_vm0_snapshot_time(self):
        """ Get UNIX timestamp of when vm-0 snapshot was last modified """

        p = subprocess.run(["lvs", "-o", "lv_name,lv_time", "--reportformat", "json"],
                           capture_output=True, check=True)

        lvs = json.loads(p.stdout.decode('utf-8'))['report'][0]['lv']
        target_lvs = list(filter(lambda x: x['lv_name'] == 'vm-0-snap', lvs))

        if len(target_lvs) > 1:
            raise RuntimeError('Found multiple lvs named vm-0-snap!')

        if len(target_lvs) == 0:
            raise RuntimeError('Failed to find LV vm-0-snap!')

        dt = datetime.datetime.strptime(target_lvs[0]['lv_time'], "%Y-%m-%d %H:%M:%S %z")
        return int(dt.timestamp())
```

Reply:
Thanks for such a detailed suggested change